File: Hard/2642. Design Graph With Shortest Path Calculator/2642. Design Graph With Shortest Path Calculator.py
from typing import List

class Graph:

    def __init__(self, n: int, edges: List[List[int]]):
        self.n = n
        self.graph = {}
        for edge in edges:
            self.addEdge(edge)

    def addEdge(self, edge: List[int]) -> None:
        from_node, to_node, edge_cost = edge
        if from_node not in self.graph:
            self.graph[from_node] = []
        self.graph[from_node].append((to_node, edge_cost))

    def shortestPath(self, node1: int, node2: int) -> int:
        if node1 == node2:
            return 0

        min_times = [float('inf')] * self.n
        min_times[node1] = 0

        visited = set()

        while len(visited) != self.n:
            source, min_time_to_source = self.get_source_with_min_time(min_times, visited)

            if min_time_to_source == float('inf'):
                return min_times[node2] if min_times[node2] != float('inf') else -1

            visited.add(source)

            for target, time_to_target in self.graph.get(source, []):
                if target in visited:
                    continue

                new_time_to_target = min_time_to_source + time_to_target
                current_time_to_target = min_times[target]

                if new_time_to_target < current_time_to_target:
                    min_times[target] = new_time_to_target

        return min_times[node2] if min_times[node2] != float('inf') else -1

    def get_source_with_min_time(self, min_times, visited):
        source = -1
        min_time = float('inf')

        for to, edge_cost in enumerate(min_times):
            if to in visited:
                continue

            if edge_cost < min_time:
                min_time = edge_cost
                source = to

        return source, min_time

File: Hard/2642. Design Graph With Shortest Path Calculator/2642. Design Graph With Shortest Path Calculator.java
import java.util.*;

public class Graph {

    private int n;
    private Map<Integer, List<int[]>> graph;

    public Graph(int n, int[][] edges) {
        this.n = n;
        this.graph = new HashMap<>();

        for (int[] edge : edges) {
            addEdge(edge);
        }
    }

    public void addEdge(int[] edge) {
        int from = edge[0];
        int to = edge[1];
        int edgeCost = edge[2];

        if (!graph.containsKey(from)) {
            graph.put(from, new ArrayList<>());
        }

        graph.get(from).add(new int[]{to, edgeCost});
    }

    public int shortestPath(int node1, int node2) {
        if (node1 == node2) {
            return 0;
        }

        int[] minTimes = new int[n];
        Arrays.fill(minTimes, Integer.MAX_VALUE);
        minTimes[node1] = 0;

        Set<Integer> visited = new HashSet<>();

        while (visited.size() != n) {
            int[] sourceWithMinTime = getSourceWithMinTime(minTimes, visited);

            int source = sourceWithMinTime[0];
            int minTimeToSource = sourceWithMinTime[1];

            if (minTimeToSource == Integer.MAX_VALUE) {
                return minTimes[node2] == Integer.MAX_VALUE ? -1 : minTimes[node2];
            }

            visited.add(source);

            if (graph.containsKey(source)) {
                for (int[] neighbor : graph.get(source)) {
                    int target = neighbor[0];
                    int timeToTarget = neighbor[1];

                    if (visited.contains(target)) {
                        continue;
                    }

                    int newTimeToTarget = minTimeToSource + timeToTarget;
                    int currentTimeToTarget = minTimes[target];

                    if (newTimeToTarget < currentTimeToTarget) {
                        minTimes[target] = newTimeToTarget;
                    }
                }
            }
        }

        return minTimes[node2] == Integer.MAX_VALUE ? -1 : minTimes[node2];
    }

    private int[] getSourceWithMinTime(int[] minTimes, Set<Integer> visited) {
        int source = -1;
        int minTime = Integer.MAX_VALUE;

        for (int to = 0; to < minTimes.length; to++) {
            if (visited.contains(to)) {
                continue;
            }

            if (minTimes[to] < minTime) {
                minTime = minTimes[to];
                source = to;
            }
        }

        return new int[]{source, minTime};
    }
}
